src/sys/redox/selector.rs

```diff
@@ -0,0 +1,206 @@
+use event::Event;
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    fs::File,
+    io::prelude::*,
+    mem,
+    os::unix::io::{AsRawFd, RawFd},
+    slice,
+    sync::{
+        Mutex,
+        atomic::{AtomicUsize, Ordering, ATOMIC_USIZE_INIT}
+    },
+    time::Duration
+};
+use syscall::{self, CLOCK_MONOTONIC, EVENT_READ, EVENT_WRITE};
+use {io, Ready, PollOpt, Token};
+
+const TIMEOUT_TOKEN: Token = Token(::std::usize::MAX - 1);
+
+/// Each Selector has a globally unique(ish) ID associated with it. This ID
+/// gets tracked by `TcpStream`, `TcpListener`, etc... when they are first
+/// registered with the `Selector`. If a type that is previously associatd with
+/// a `Selector` attempts to register itself with a different `Selector`, the
+/// operation will return with an error. This matches windows behavior.
+static NEXT_ID: AtomicUsize = ATOMIC_USIZE_INIT;
+
+#[derive(Debug)]
+pub struct Selector {
+    id: usize,
+    efd: File,
+    tokens: Mutex<BTreeMap<RawFd, BTreeSet<Token>>>
+}
+
+impl Selector {
+    pub fn new() -> io::Result<Selector> {
+        let efd = File::open("event:")?;
+
+        // offset by 1 to avoid choosing 0 as the id of a selector
+        let id = NEXT_ID.fetch_add(1, Ordering::Relaxed) + 1;
+
+        Ok(Selector {
+            id: id,
+            efd: efd,
+            tokens: Mutex::new(BTreeMap::new()),
+        })
+    }
+
+    pub fn id(&self) -> usize {
+        self.id
+    }
+
+    /// Wait for events from the OS
+    pub fn select(&self, evts: &mut Events, awakener: Token, timeout: Option<Duration>) -> io::Result<bool> {
+        let mut timeout_fd = None;
+        if let Some(timeout) = timeout {
+            let mut file = File::open(format!("time:{}", CLOCK_MONOTONIC))?;
```

Comment:
Do you want to allocate a String everytime select is called?

This could be resolved in a future pr.

Reply:
Good point, this is static and could be cached. Optimally would be if Rust would just allow `concat!`ing with constants :(

Reply:
Although, fun fact, `File::open` allocates on each occurance on linux. This is because it needs a terminating nul byte. So this isn't that expensive

Reply:
Didn't about `File::open` allocating. Then lets leave this for a future pr.